Refactor break up stock

jambroo opened the pull request:
This is a series of simple changes to refactor server/controllers/stock/reports/index.js, which was requested in issue #2068.

Closes #2068.

Commit:
moving common function and lots report functionality into separate files.

Commit:
add more functionality to common file and clean up stock lots report.

Commit:
stock adjustment receipt refactored.

Commit:
entry depot receipt refactored.

Commit:
entry integration receipt refactored.

Commit:
stock entry purchase receipt refactored.

Commit:
stock exit depot receipt refactored.

Commit:
stock exit loss receipt refactored.

Commit:
stock exit patent receipt refactored.

Commit:
stock exit sevice receipt refactored.

Commit:
stock exit inventories refactored.

Commit:
stock exit inventory refactored.

Commit:
stock movement report refactored.

Commit:
fixing invalid export.

jniles committed:
Merge branch 'master' into refactor-break-up-stock

jniles commented on server/controllers/stock/reports/stock/inventory_report.js:
```diff
@@ -0,0 +1,59 @@
+    .done();
+}
+
+exports.stockInventoyReport = stockInventoriesReport;
```
This is where you are running into errors - this should be `stockInventoryReport`, not `stockInventoriesReport`.

We are a multilingual team and sometimes have a few confusing variable names. :)

jniles commented on server/controllers/stock/reports/stock/inventory_report.js:
```diff
@@ -0,0 +1,59 @@
+
+  const optionReport = _.extend(req.query, {
+    filename : 'TREE.STOCK_INVENTORY_REPORT',
+    orientation : 'landscape',
```
It looks like almost every report uses the following object:
```js
{
  orientation : 'landscape',
  footerRight : '[page] / [toPage]',
  footerFontSize : '8',
}
```

Could you put this object in `common.js`?  Then it will be available to use with `_.extend()`, [which can take multiple objects](https://lodash.com/docs/4.17.4#assignIn).  It would look something like this:
```js
const optionReports = _.extend(req.query, pdfOptions, {
  filename : 'TREE.STOCK_INVENTORY_REPORT',
});
```

jniles reviewed (commented):
This is looking really good!  Great work.  :+1:

I think I've pinpointed the source of the trouble and made a subsequent recommendation for refactoring.

Let me know if you need any help!

Commit:
fixing export variable typo.

Commit:
pdf options put in common file.

jniles commented on server/controllers/stock/reports/index.js:
```diff
@@ -7,868 +7,18 @@
+const stockMovementsReport = require('./stock/movements_report');
+const stockInventoriesReport = require('./stock/inventories_report');
+const stockInventoryReport = require('./stock/inventory_report');
+const stockExitPatientReceipt = require('./stock/exit_patient_receipt');
```
It looks like you have a typo in the file name which makes this `require()` fail.  The file is called `exit_patent_receipt` instead of `exit_patient_receipt`.

jniles replied:
Sweet!   This means that there is an `app.get()` somewhere that isn't being given a function.  We'll have to track that down...

jniles replied:
Nothing looks wrong online - let me pull your branch and play around with it.  I'll report back when I'm able to figure something out!

jniles replied:
Aha - gotcha.  Each of the `require()`s are nested one too deep.  In other words, you export the functions using `exports.fn = fn`, and then pass them up to the top via another `exports.fn = fn`.  Specifically, each of [these lines](https://github.com/IMA-WorldHealth/bhima-2.X/pull/2084/files#diff-a51ce60ba2265ea47425f240a42b96cdL874) should either read:
```js
 exports.stockLotsReport = stockLotsReport.stockLotsReport;
// etc ...
```
OR

You can change [the exports line to using module.exports](https://github.com/IMA-WorldHealth/bhima-2.X/pull/2084/files#diff-37cc91ea7119fcc956060bf5ae71be36R70).

I've gone ahead and fixed this while I was messing around with you code base.  I chose the `module.exports` pattern.  I think it should all pass - let's see if Travis is happy with it!

jniles reviewed (commented):
Changes are looking very good!

To debug the tests, you can take a look at the Travis log.  For example, I deduced the filename error from [this line failing](https://travis-ci.org/IMA-WorldHealth/bhima-2.X/jobs/273240971#L861) in our test suite.  Unfortunately, the tests take a while to run because we lump everything together ... we'll be trying to optimize our opts in the future, but it isn't a priority currently.

Thanks for all your help!

Commit:
typo fix.

jniles committed:
fix(stock reports): use module.exports for exposure

This commit fixes the stock reports links that use one to many
`exports.fn` resulting in broken links.  To avoid changing the API, I've
refactored the code to use `module.exports` and only export the single
function.

jniles commented on server/controllers/stock/reports/stock/exit_service_receipt.js:
```diff
@@ -0,0 +1,69 @@
+const { _, ReportManager, Stock, identifiers, NotFound, STOCK_EXIT_SERVICE_TEMPLATE } = require('../common');
```
Aha!  The bug is here - `db` is not defined, and should be imported.

This kind of error can be caught fairly easily if you have [eslint](https://eslint.org/) installed in your editor.  It has a ton of [integrations](https://eslint.org/docs/user-guide/integrations), so your preferred editor is probably supported.

Guess that's why we have tests ;)

Commit:
fixed import newline and added missing db import for exit service report.

jniles reviewed (approved):
Awesome!  Code LGTM.
